src/bsoncxx/include/bsoncxx/v_noabi/bsoncxx/stdx/iterator.hpp

```diff
@@ -0,0 +1,227 @@
+#pragma once
+
+#include <initializer_list>
+#include <iterator>
+#include <memory>
+#include <type_traits>
+#include <utility>
+#include <vector>
+#if defined(__has_include)
+#if __has_include(<version>)
+#include <version>
+#endif
+#endif
+
+#include "./operators.hpp"
+#include "./type_traits.hpp"
+
+#include <bsoncxx/config/prelude.hpp>
+
+namespace bsoncxx {
+inline namespace v_noabi {
+namespace detail {
+
+/**
+ * @brief Backport of std::pointer_traits with the addition of to_address() for pointers
+ */
+template <typename T>
+struct pointer_traits : std::pointer_traits<T> {};
+
+template <typename T>
+struct pointer_traits<T*> : std::pointer_traits<T*> {
+    using typename std::pointer_traits<T*>::pointer;
+    using typename std::pointer_traits<T*>::element_type;
+
+    static constexpr pointer to_address(pointer p) noexcept {
+        return p;
+    }
+};
+
+struct _to_address_fn {
+    template <typename FancyPointer>
+    static constexpr auto impl(FancyPointer fp, rank<3>)
+        bsoncxx_returns(pointer_traits<FancyPointer>::to_address(fp));
+
+    template <typename Iterator>
+    static constexpr auto impl(Iterator iter, rank<2>) bsoncxx_returns(iter.operator->());
+
+    template <typename T>
+    static constexpr auto impl(T* p, rank<1>) bsoncxx_returns(p);
+
+    template <typename Iterator>
+    constexpr auto operator()(Iterator iter) const bsoncxx_returns((impl)(iter, rank<10>{}));
+};
+
+/**
+ * @brief Convert an iterator or pointer-like type to a raw pointer.
+ */
+constexpr static _to_address_fn to_address;
+
+/**
+ * @brief The type obtained by to_address() for the given object, if valid
+ */
+template <typename T>
+using to_address_t = decltype(to_address(std::declval<T&>()));
+
+/**
+ * @brief The result of applying unary operator* to the given object, if valid
+ */
+bsoncxx_push_warnings();
+bsoncxx_disable_warning(Clang("-Wvoid-ptr-dereference"));
+template <typename I>
+using dereference_t = decltype(*std::declval<I>());
+bsoncxx_pop_warnings();
+
+/**
+ * @brief Detect a type that can be dereferenced (like a pointer) and the result
+ * type is non-void
+ */
+template <typename I>
+struct is_dereferencable : conjunction<is_detected<dereference_t, I>,
+                                       // Clang supports dereferencing void*, and we can't detect
+                                       // that easily. Refuse if I is (cv-)void*
+                                       negation<std::is_void<remove_pointer_t<I>>>,
+                                       negation<std::is_void<detected_t<dereference_t, I>>>> {};
```

Comment:
Possibly, but any other use of `dereference_t` (including the upcoming optional<T> changes) that could potentially trigger the warning, so I think it would be easier to just supress the warning in one location.